File: frontend/web/src/types/customization.ts
export interface DashboardWidgetConfig {
  key: string
  title: string
  type: string
}

export interface DashboardPreferences {
  userId: string
  dashboardTemplateId?: string | null
  widgets: DashboardWidgetConfig[]
  layout: Record<string, unknown>
  updatedAt: string
}

export interface DashboardTemplate {
  id: string
  name: string
  description?: string | null
  widgetsJson: Record<string, unknown>
  layoutJson: Record<string, unknown>
  isDefault: boolean
}

export interface SavedReportWidget {
  key: string
  title: string
  type: string
}

export interface SavedReport {
  id: string
  name: string
  description?: string | null
  reportType: string
  filtersJson: Record<string, unknown>
  widgetConfigJson: SavedReportWidget[]
  createdBy: string
  createdAt: string
  updatedAt: string
}

export interface ReportWidgetConfig {
  id: string
  name: string
  key: string
  category: string
  description?: string | null
  configJson: Record<string, unknown>
}

File: backend/schemas/customization.py
from __future__ import annotations

from datetime import datetime
from typing import Optional

from schemas.common import CamelBase


class DashboardWidgetConfig(CamelBase):
    code: str
    visible: bool = True
    order: int = 0
    title: Optional[str] = None
    filters: dict = {}


class DashboardPreferencesResponse(CamelBase):
    user_id: str
    dashboard_template_id: Optional[str] = None
    widgets: list[DashboardWidgetConfig] = []
    layout: dict = {}
    updated_at: datetime


class DashboardPreferencesUpdate(CamelBase):
    dashboard_template_id: Optional[str] = None
    widgets: list[DashboardWidgetConfig] = []
    layout: dict = {}


class DashboardTemplateResponse(CamelBase):
    id: str
    name: str
    description: Optional[str] = None
    config_json: dict


class SavedReportResponse(CamelBase):
    id: str
    user_id: str
    name: str
    report_type: str
    filters_json: dict
    config_json: dict
    created_at: datetime
    updated_at: datetime


class SavedReportCreate(CamelBase):
    name: str
    report_type: str
    filters_json: dict = {}
    config_json: dict = {}


class ReportWidgetConfigResponse(CamelBase):
    id: str
    code: str
    name: str
    category: str
    config_json: dict
